### PR TITLE
SONARIAC-772 Rule S6385: ARM Azure custom roles should not grant subscription Owner capabilities

### PR DESCRIPTION
**Pull request by petertrr**

### CODE REVIEW

File: iac-extensions/arm/src/main/java/org/sonar/iac/arm/checks/SubscriptionOwnerCapabilitiesCheck.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  private static Stream<Expression> findArbitraryActions(ContextualObject permissionObject) {
+    return permissionObject.list("actions").getItemIf(
+      action -> isValue(action, "*").isTrue());
+  }
+
+  private static boolean isSensitiveScope(Tree scope) {
+    boolean hasSensitiveCall = isValue(scope, "[managementGroup().id]").isTrue() ||
+      isValue(scope, "[subscription().id]").isTrue();
+    boolean referencesSensitiveScope = matchesValue(scope, s -> s.startsWith("/subscriptions/")).isTrue() &&
+      matchesValue(scope, s -> RESOURCE_GROUP_SCOPE_PATTERN.matcher(s).matches()).isFalse() ||
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/arm/src/main/java/org/sonar/iac/arm/checks/SubscriptionOwnerCapabilitiesCheck.java:83]: I would use the same approach as for the terraform rule, with this regex: `^/subscriptions/[^/]+/?$`
That way you don't need to perform two checks on the same string.

File: iac-extensions/arm/src/test/resources/checks/SubscriptionOwnerCapabilitiesCheck/Microsoft.Authorization_roleDefinitions/test.json [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      "name": "Compliant - specific actions are specified",
+      "properties": {
+        "permissions": [
+          {
+            "actions": ["Microsoft.Compute/*"],
+            "notActions": []
+          }
+        ],
+        "assignableScopes": [
+          "[subscription().id]"
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/arm/src/test/resources/checks/SubscriptionOwnerCapabilitiesCheck/Microsoft.Authorization_roleDefinitions/test.json:17]:
```suggestion
          "${type}"
```
You could replace this with the injected sensitive value directly.

File: iac-extensions/arm/src/test/resources/checks/SubscriptionOwnerCapabilitiesCheck/Microsoft.Authorization_roleDefinitions/test.json [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+      "name": "Compliant - scope is limited to a resource group",
+      "properties": {
+        "permissions": [
+          {
+            "actions": ["*"],
+            "notActions": []
+          }
+        ],
+        "assignableScopes": [
+          "/subscriptions/b24988ac-6180-42a0-ab88-20f7382dd24c/resourceGroups/b24988ac-6180-42a0-ab88-20f7382dd24c"
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/arm/src/test/resources/checks/SubscriptionOwnerCapabilitiesCheck/Microsoft.Authorization_roleDefinitions/test.json:33]:
```suggestion
          "${type}"
```
Same here.

**Review by rudy-regazzoni-sonarsource — APPROVED** [review]
LGTM! Please see my comments.

File: iac-extensions/arm/src/test/resources/checks/SubscriptionOwnerCapabilitiesCheck/Microsoft.Authorization_roleDefinitions/test.json (hunk 3 above)
Comment (petertrr) [reply iac-extensions/arm/src/test/resources/checks/SubscriptionOwnerCapabilitiesCheck/Microsoft.Authorization_roleDefinitions/test.json:33]: Here that's a compliant example with insensitive scope, but the test is parameterized with different types of sensitive scopes.

**Pull request merged**